service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java

```diff
@@ -2041,49 +2041,71 @@ public void saveAdminCommentToOrder(AdminCommentDto adminCommentDto, String uuid
-                    collectedValue.append(
-                        collectInfoAboutEcoNumberEventHistory(i, oldNumber, ecoNumber.getNewEcoNumber()));
-                }
+        Set<String> oldEcoNumbers = Set.copyOf(order.getAdditionalOrders());
+        Set<String> newEcoNumbers = ecoNumberDto.getEcoNumber();
+
+        int countOfNewNumbers = 0;
+        int countOfDeleted = 0;
+
+        StringBuilder collectedNewValue = new StringBuilder();
+        StringBuilder collectedDeletedValue = new StringBuilder();
+        for (String newNumber : newEcoNumbers) {
+            if (!oldEcoNumbers.contains(newNumber)
+                && !newNumber.equals("")) {
+                order.getAdditionalOrders().add(newNumber);
+                collectedNewValue.append(collectInfoAboutAddNewEcoNumber(countOfNewNumbers, newNumber));
+                countOfNewNumbers++;
             }
-            if (collectedValue.length() > 1) {
-                eventService.save(collectedValue.toString(),
-                    currentUser.getRecipientName() + "  " + currentUser.getRecipientSurname(), order);
+        }
+        for (String oldNumber : oldEcoNumbers) {
+            if (!newEcoNumbers.contains(oldNumber)) {
+                order.getAdditionalOrders().remove(oldNumber);
+                collectedDeletedValue.append(collectInfoAboutDeletedEcoNumber(countOfDeleted, oldNumber));
+                countOfDeleted++;
             }
         }
+
+        orderRepository.save(order);
+        eventService.save(allHistoryChanges(collectedNewValue, collectedDeletedValue),
+            currentUser.getRecipientName() + "  " + currentUser.getRecipientSurname(), order);
     }
 
-    /**
-     * This is method which collects info about eco number for event history.
-     *
-     * @author Yuriy Bahlay.
-     */
-    private String collectInfoAboutEcoNumberEventHistory(int i, String oldNumber, String newEcoNumber) {
+    private String allHistoryChanges(
+        StringBuilder collectedNewValue,
+        StringBuilder collectedDeletedValue) {
+        StringBuilder finalHistory = new StringBuilder();
+        if (!collectedNewValue.toString().isEmpty()) {
+            finalHistory.append(collectedNewValue);
+        }
+        if (!collectedDeletedValue.toString().isEmpty()) {
+            finalHistory.append(collectedDeletedValue);
+        }
+        return finalHistory.toString();
+    }
+
+    private String collectInfoAboutAddNewEcoNumber(int i, String newEcoNumber) {
         StringBuilder values = new StringBuilder();
         if (i == 0) {
-            values.append(OrderHistory.CHANGES_ECO_NUMBER);
+            values.append(OrderHistory.ADD_NEW_ECO_NUMBER + ": ");
         }
-        if (i > 0) {
-            values.append(";");
+        values.append(newEcoNumber + "; ");
+        return values.toString();
+    }
+
+    private String collectInfoAboutDeletedEcoNumber(int i, String oldNumber) {
```

Comment:
collectInfoAboutAddNewEcoNumber and collectInfoAboutDeletedEcoNumber do the same thing. you can create new method with extra argument with type `OrderHistory`

Reply:
Fixed